### PR TITLE
Public Proxy Address

### PR DESCRIPTION
**Pull request by russjones**
**Purpose**

At the moment when running `teleport status` we report the address of the first proxy that has heartbeated back to the AuthServer. If you have multiple proxies or if you have a defined address that you would like reported, Teleport does not support this functionality.

This PR fixes this by introducing the `public_address` field under `proxy_service` in `teleport.yaml`. When set, this is the address that will be returned whenever `teleport status` is run.

**Implementation**

* Added `public_address ` to `proxy_service` in `lib/config/*`.
* Added `proxyPublicAddr` to `lib/srv/sshserver.go` so it's accessible to all proxies and nodes.
* Added `public_addr` field to `Server` resource.
* In `registerServer` in `lib/srv/sshserver.go` if the server is type proxy, the public address is passed along.
* Updated the value set to `SSH_SESSION_WEBPROXY_ADDR` to be be the proxy public address and if it doesn't exist, the hostname.

### CODE REVIEW

File: lib/config/fileconf.go [hunk 1]
```diff
@@ -479,11 +480,12 @@ type CommandLabel struct {
...
-	WebAddr  string `yaml:"web_listen_addr,omitempty"`
-	TunAddr  string `yaml:"tunnel_listen_addr,omitempty"`
-	KeyFile  string `yaml:"https_key_file,omitempty"`
-	CertFile string `yaml:"https_cert_file,omitempty"`
+	Service    `yaml:",inline"`
+	WebAddr    string `yaml:"web_listen_addr,omitempty"`
+	TunAddr    string `yaml:"tunnel_listen_addr,omitempty"`
+	KeyFile    string `yaml:"https_key_file,omitempty"`
+	CertFile   string `yaml:"https_cert_file,omitempty"`
+	PublicAddr string `yaml:"public_address",omitempty`
```
Comment (klizhentas) [inline lib/config/fileconf.go]: `public_addr` to be consistent with `web_listen_addr` above

File: lib/srv/sshserver.go [hunk 2]
```diff
@@ -282,14 +285,23 @@ func (s *Server) getAdvertiseIP() net.IP {
...
 	if s.getAdvertiseIP() == nil {
 		return s.addr.Addr
 	}
 	_, port, _ := net.SplitHostPort(s.addr.Addr)
 	return net.JoinHostPort(s.getAdvertiseIP().String(), port)
 }
 
+// ProxyPublicAddr returns the public address the Web UI can be accessed at
+// as set in Teleport configuration. If it is not set, an error is returned.
+func (s *Server) ProxyPublicAddr() (string, error) {
```
Comment (klizhentas) [inline lib/srv/sshserver.go]: I would just return empty string, as it's not an error.

File: lib/srv/exec.go [hunk 3]
```diff
@@ -165,20 +165,27 @@ func prepareCommand(ctx *ctx) (*exec.Cmd, error) {
...
-	// session URL. fall back to <proxyhost> placeholder
-	proxyHost := "<proxyhost>"
+
+	// try to get the address the proxy server can be accessed at. if public_address
+	// is set in the configuration use that, if not try to fall back to the 1st
+	// available proxy from the auth server heartbeat.
+	proxyHost := "<proxyhost>:3080"
 	if ctx.srv != nil {
-		proxies, err := ctx.srv.authService.GetProxies()
+		proxyHost, err = ctx.srv.ProxyPublicAddr()
```
Comment (klizhentas) [inline lib/srv/exec.go]: this will never be true for server, because exec is only available on proxies. This will always return error.

**Review by klizhentas — CHANGES_REQUESTED** [review]
In addition to the comments, can you do the following:

* Add public_addr to proxy resource
* Use it when requesting  list or proxies
* Add the address to the heartbeat.

**Comment by russjones** [comment]
@klizhentas Made the changes you suggested, can you take another look?

**Review by klizhentas — APPROVED** [review]

File: lib/srv/exec.go [hunk 4]
```diff
@@ -165,20 +165,27 @@ func prepareCommand(ctx *ctx) (*exec.Cmd, error) {
...
-			proxyHost = proxies[0].GetHostname()
+			proxies, err := ctx.srv.authService.GetProxies()
+			if err != nil {
+				log.Error(err)
+			}
+			if len(proxies) > 0 {
+				proxyHost = proxies[0].GetHostname()
+			}
+
+			log.Infof("Proxy public address not set in configuration, defaulting to first proxy from heartbeat: %q", proxyHost)
```
Comment (russjones) [inline lib/srv/exec.go]: Weird, I think this was still showing an outdated commit. It's `log.Debugf` now.

**Pull request merged**